File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/ruby/stack.rb
class Stack
  attr_reader :limit

  def initialize
    @stack = []
    # this is an arbitrary (low) value to make testing easier
    @limit = 10
  end

  # add item to top of stack if not full
  # if full, throw error
  def push(item)
    raise "Stack is full" if isFull?
    @stack.push(item)
  end

  # remove item from top of stack and return it
  def pop
    @stack.pop
  end

  # return item at top of stack without removing it
  def peek
    @stack.last
  end

  # return true if stack is empty, otherwise false
  def isEmpty?
    size.zero?
  end

  # return true if stack is full, otherwise false
  def isFull?
    size === @limit
  end

  # return number of items in stack
  def size
    @stack.length
  end

  # return -1 if item not in stack, 
  # otherwise integer representing how far it is from the top (array length - index -1)
  def search(target)
    @stack.each_with_index do |item, idx|
      return size - idx - 1 if item === target
    end
    return -1
  end

  # print contents of stack: do not return the stack itself!
  def print
    puts @stack.join('-')
  end

end

if __FILE__ == $PROGRAM_NAME
  stack = Stack.new
 
  (0...6).each do |num|
    stack.push(num)
  end

  puts "size: #{stack.size}"
  puts "is empty?: #{stack.isEmpty?}"
  puts "is full?: #{stack.isFull?}"
  puts "find 3: #{stack.search(3)}"
  puts "peek: #{stack.peek}"

  stack.print
end

File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/javascript/stack.js
class Stack {
  constructor() {
    this.stack = [];
    // this is an arbitrary (low) value to make testing easier
    this.limit = 10;
  }

  // add item to top of stack if not full
  // if full throw error
  push(item) {
    if(this.isFull()) {
      throw new Error ("Stack is Full");
    } else {
      this.stack.push(item);
    };
  }

  // remove item from top of stack and return it
  pop() {
    return this.stack.pop();
  }

  // return item at top of stack without removing it
  peek() {
    return this.stack[this.size() -1];
  }

  // return true if stack is empty, otherwise false
  isEmpty() {
    return this.size() === 0;
  }

  // return true if stack is full, otherwise false
  isFull() {
    return this.size() === this.limit;
  }

  // return number of items in stack
  size() {
    return this.stack.length;
  }

  // return -1 if item not in stack, otherwise integer representing 
  // how far it is from the top (top = 0, so stack length - index - 1)
  search(target) {
    for(let i = 0; i < this.size(); i++) {
      if(this.stack[i] === target) {
        return this.size() - i - 1;
      };
    };
    return -1;
  }

  // print contents of stack: do not return the stack itself!
  print() {
    console.log(this.stack.join('-'));
  }
}

if (require.main === module) {
  const stack = new Stack();

  for(let i = 0; i < 6; ++i) {
    stack.push(i);
  }

  console.log('Print');
  stack.print();
  console.log('Size', stack.size());
  console.log('Is Empty?', stack.isEmpty());
  console.log('Is Full?', stack.isFull());
  console.log('Peek', stack.peek());
  console.log('Find 6', stack.search(6));
}

module.exports = Stack;
